### ISSUE TITLE
Hi! Why didn't train for self._discriminator?

### ISSUE THREAD

**Issue by yonghuixu** [zurutech/ashpy#20]
real_x, real_y = real_xy

        if len(self._generator.inputs) == 2:
            g_inputs = [g_inputs, real_y]

        with tf.GradientTape(persistent=True) as tape:
            fake = self._generator(g_inputs, training=True)
            logits_fake = self._discriminator(fake)  #why didn't add this and next line?
            logits_real = self._discriminator(real_y)
            d_loss = self._d_loss(
                self._context, fake=fake, real=real_x, condition=real_y, training=True
            )

            g_loss = self._g_loss(
                self._context, fake=fake, real=real_x, condition=real_y, training=True
            )

        # check that we have some trainable_variables
        assert self._generator.trainable_variables
        assert self._discriminator.trainable_variables

        # calculate the gradient
        d_gradients = tape.gradient(d_loss, self._discriminator.trainable_variables)
        g_gradients = tape.gradient(g_loss, self._generator.trainable_variables)

**Comment by galeone** [comment]
Hi @YonghuiXu and thank you for your interest in AshPy!

The computation of the `logits_fake` and `logits_real` is delegated to the losses computation.

The AshPy losses are powerful concepts, in fact, these are not simple loss functions but are Executors (see https://github.com/zurutech/ashpy/blob/master/ashpy/losses/executor.py).

An executor is a concept that allows the loss object to having a context to use; the context contains everything the loss needs to be executed.

Thus, the `logits_fake` and `logits_real` computation are not executed inside the trainer but is the Executor object that computes them (if it needs them).

Have a look at the [AdversarialLossD](https://github.com/zurutech/ashpy/blob/master/ashpy/losses/gan.py#L531) to see how the adversarial training of the discriminator is performed, and also look at [AdversarialLossG](https://github.com/zurutech/ashpy/blob/master/ashpy/losses/gan.py#L114) to see the adversarial trainer of the Generator (have a look at the `call` method).

We designed the trainer in this way, in order to have only a single training loop (that invokes `self._d_loss` and `self._g_loss`) and be able to customize the training only by changing the Executors.

If you have any other question feel free to ask them here!

**Comment by galeone** [comment]
Closing for inactivity.
